container/shim/src/index.js

```diff
@@ -42,159 +48,246 @@ const PROXY_RESPONSE_HEADERS = [
+  const controller = new AbortController()
+  const timeout = setTimeout(() => {
+    controller.abort()
+  }, GATEWAY_TIMEOUT)
+
+  const ipfsReq = https.get(ipfsUrl, {
+    agent: ipfsAgent,
+    timeout: GATEWAY_TIMEOUT,
+    headers: { 'User-Agent': NODE_UA },
+    signal: controller.signal
+  }, async fetchRes => {
+    clearTimeout(timeout)
+    const { statusCode } = fetchRes
+    if (statusCode >= 400) {
+      debug.extend('error')(`Invalid response from IPFS gateway (${statusCode}) for ${cid}`)
     }
 
-    const ipfsUrl = new URL(IPFS_GATEWAY_ORIGIN + req.path)
-    if (format) {
-      ipfsUrl.searchParams.set('format', format)
+    res.status(statusCode)
+    proxyResponseHeaders(fetchRes, res)
+
+    if (format === 'car') {
+      streamCAR(fetchRes, res).catch(() => {})
+    } else {
+      fetchRes.pipe(res)
     }
-    for (const key of ['filename', 'download']) {
-      if (key in req.query) {
-        ipfsUrl.searchParams.set(key, req.query[key])
-      }
+  }).on('error', err => {
+    clearTimeout(timeout)
+    debug.extend('error')(`Error fetching from IPFS gateway for ${cid}: ${err.name} ${err.message}`)
+    if (controller.signal.aborted) {
+      return res.sendStatus(504)
+    }
+    res.sendStatus(502)
+  }).on('timeout', () => {
+    clearTimeout(timeout)
+    debug.extend('error')(`Timeout from IPFS gateway for ${cid}`)
+    ipfsReq.destroy()
+    res.destroy()
+  })
+
+  req.on('close', () => {
+    clearTimeout(timeout)
+    if (!res.writableEnded) {
+      debug.extend('error')('Client aborted early, terminating gateway request')
+      ipfsReq.destroy()
     }
+  })
+}
 
-    const controller = new AbortController()
-    const timeout = setTimeout(() => {
-      controller.abort()
-    }, GATEWAY_TIMEOUT)
-
-    const ipfsReq = https.get(ipfsUrl, {
-      agent: ipfsAgent,
-      timeout: GATEWAY_TIMEOUT,
-      headers: { 'User-Agent': NODE_UA },
-      signal: controller.signal
-    }, async fetchRes => {
-      clearTimeout(timeout)
-      const { statusCode } = fetchRes
-      if (statusCode >= 400) {
-        debug.extend('error')(`Invalid response from IPFS gateway (${statusCode}) for ${cid}`)
+async function maybeRespondFromL2 (req, res, { cid, format }) {
+  debug(`Fetch ${req.path} from L2s`)
+  const cidHash = crypto.createHash('sha512').update(cid).digest()
+  Array.from(connectedL2Nodes.values())
+    .map(l2Node => ({
+      ...l2Node,
+      distance: xorDistance(
+        cidHash,
+        l2Node.idHash
+      )
+    }))
+    .sort((a, b) => xorDistance.compare(a.distance, b.distance))
+    .slice(0, 3)
+    .forEach(({ res }) => {
+      const payload = {
+        requestId: req.get('saturn-transfer-id'),
+        cid
       }
+      res.write(`${JSON.stringify(payload)}\n`)
+    })
+  if (L2_FIRE_AND_FORGET) {
+    return false
+  }
 
-      res.status(statusCode)
-      proxyResponseHeaders(fetchRes, res)
+  const onResponse = pDefer()
+  openCARRequests.set(cid, onResponse)
```

Comment:
The logic only handles cids, I think we also need to handle file paths in the L2 request payload, cache keys (`openCARRequests`), endpoints (`app.post('/data/:cid'`), etc.

Currently if I request https://strn.pl/ipfs/QmWKDHujv2nNZkBaxj7UvuraoVzgH79pbZqxGyy89e4QnM/999.json, I'd receive https://strn.pl/ipfs/QmWKDHujv2nNZkBaxj7UvuraoVzgH79pbZqxGyy89e4QnM instead. The former is a file, the latter is a directory.

Reply:
Good catch!

Reply:
Asked for clarification in https://filecoinproject.slack.com/archives/C031K3BG527/p1661250640769579

Reply:
as discussed, I've disabled l2 fallback